presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DistinctAggregationToGroupBy.java

```diff
@@ -0,0 +1,457 @@
+ *
+ *  SELECT a1, a2,..., an, F1(b1), F2(b2), F3(b3), ...., Fm(bm), F1(distinct c1), ...., Fm(distinct cm) FROM Table GROUP BY a1, a2, ..., an
+ *
+ *  INTO
+ *
+ *  SELECT a1, a2,..., an, arbitrary(if(group = 0, f1)),...., arbitrary(if(group = 0, fm)), F(if(group = 1, c1)), ...., F(if(group = m, cm)) FROM
+ *      SELECT a1, a2,..., an, F1(b1) as f1, F2(b2) as f2,...., Fm(bm) as fm, c1,..., cm group FROM
+ *        SELECT a1, a2,..., an, b1, b2, ... ,bn, c1,..., cm FROM Table GROUP BY GROUPING SETS ((a1, a2,..., an, b1, b2, ... ,bn), (a1, a2,..., an, c1), ..., ((a1, a2,..., an, cm)))
+ *      GROUP BY a1, a2,..., an, c1,..., cm group
+ *  GROUP BY a1, a2,..., an
+ */
+
+public class DistinctAggregationToGroupBy
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(DistinctAggregationToGroupBy::hasDistinctInput)
+            .matching(DistinctAggregationToGroupBy::hasSingleDistinctArgument)
+            .matching(DistinctAggregationToGroupBy::noFilters)
+            .matching(DistinctAggregationToGroupBy::noMasks)
+            .matching(DistinctAggregationToGroupBy::noOrdering);
+
+    private static boolean hasDistinctInput(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .map(Aggregation::getCall)
+                .filter(FunctionCall::isDistinct)
+                .map(FunctionCall::getArguments)
+                .<Set<Expression>>map(HashSet::new)
+                .distinct()
+                .count() > 0;
+    }
+
+    private static boolean hasSingleDistinctArgument(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .map(Aggregation::getCall)
+                .filter(FunctionCall::isDistinct)
+                .allMatch(c -> c.getArguments().size() == 1);
+    }
+
+    private static boolean noFilters(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .map(Aggregation::getCall)
+                .noneMatch(call -> call.getFilter().isPresent());
+    }
+
+    private static boolean noMasks(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .noneMatch(e -> e.getMask().isPresent());
+    }
+
+    private static boolean noOrdering(AggregationNode aggregation)
+    {
+        return !aggregation.hasOrderings();
+    }
+
+    private final Metadata metadata;
+
+    public DistinctAggregationToGroupBy(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode node, Captures captures, Context context)
```

Comment:
I find this implementation hard to read. I think it would help if you structured it such that it builds things from the bottom up and return objects that contain the node that was created + anything that's needed to build the node above it. However, I haven't tried playing this out, so not sure how much cleaner it will make it. You'll have to give it a try and see how it goes.